airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py

```diff
@@ -0,0 +1,129 @@
+
+        WARNING: This is used to retain backwards compatibility with streams using the legacy get_stream_state() method.
+        In order to uptake Resumable Full Refresh, connectors must migrate streams to use the state setter/getter methods.
+        """
+        # todo blai: Ideally observe and get_checkpoint should just be one method, but because of the legacy state behavior
+        #  observation and reading Stream.state checkpoint are not 1:1 with each other
+
+    @abstractmethod
+    def get_checkpoint(self) -> Optional[Mapping[str, Any]]:
+        """
+        Retrieves the current state value of the stream
+        """
+
+
+class IncrementalCheckpointReader(CheckpointReader):
+    """
+    IncrementalCheckpointReader handles iterating through a stream based on partitioned windows of data that are determined
+    before syncing data.
+    """
+
+    def __init__(self, stream_state: Mapping[str, Any], stream_slices: Iterable[Optional[Mapping[str, Any]]]):
+        self._state: Optional[Mapping[str, Any]] = stream_state
+        self._stream_slices = iter(stream_slices)
+        self._has_slices = False
+
+    def next(self) -> Optional[Mapping[str, Any]]:
+        try:
+            next_slice = next(self._stream_slices)
+            self._has_slices = True
+            return next_slice
+        except StopIteration:
+            # This is used to avoid sending a duplicate state message at the end of a sync since the stream has already
+            # emitted state at the end of each slice. If we want to avoid this extra complexity, we can also just accept
+            # that every sync emits a final duplicate state
+            if self._has_slices:
+                self._state = None
+            return None
+
+    def observe(self, new_state: Mapping[str, Any]) -> None:
+        self._state = new_state
+
+    def get_checkpoint(self) -> Optional[Mapping[str, Any]]:
+        return self._state
+
+
+class ResumableFullRefreshCheckpointReader(CheckpointReader):
+    """
+    ResumableFullRefreshCheckpointReader allows for iteration over an unbounded set of records based on the pagination strategy
+    of the stream. Because the number of pages is unknown, the stream's current state is used to determine whether to continue
+    fetching more pages or stopping the sync.
+    """
+
+    def __init__(self, stream_state: Mapping[str, Any]):
+        # The first attempt of an RFR stream has an empty {} incoming state, but should still make a first attempt to read records
+        # from the first page in next().
+        self._first_page = bool(stream_state == {})
+        self._state: Mapping[str, Any] = stream_state
+
+    def next(self) -> Optional[Mapping[str, Any]]:
+        #  todo blai: I think this is my main concern with the interface is that it puts a lot of onus on the connector developer to
+        #   structure their state object correctly to coincide with how the checkpoint_reader interprets values.
```

Comment:
I've talked about this aspect at length with a few other engineers. I really don't know a better way forward in the immediate right now. Having it be non-obvious that the connector developer be state aware is not a good DX.

The longer term solution would either be a structured state class that handles how to read/write state using clearly defined method instead of a generic map. Or through a cursor class. The problem is it doesn't current exist in the legacy Python CDK

However, we can potentially rationalize that this is okay in the immediate for two reasons:
- This should be less of a problem for low-code because the actual state management code is owned by us and not the connector developer. It also already supports cursors
- We also want to start moving more things to concurrent where we already have a concept of cursors and is worth integrating there as well.

Basically, even if this interface is not a good paradigm shift in the short term, if we're focusing on concurrent + low-code long term, we can accept this. Happy to be challenged on this as well